db/snapshot/backup_task.cc

```diff
@@ -135,24 +148,37 @@ future<> backup_task_impl::backup_file(const sstring& name) {
     co_await coroutine::maybe_yield();
 };
 
+future<> backup_task_impl::backup_sstable(sstables::generation_type gen, const comps_vector& comps) {
+    snap_log.debug("Backing up SSTable generation {}", gen);
+    for (auto it = comps.begin(); it != comps.end() && !_ex; ++it) {
+        // Pre-upload break point. For testing abort in actual s3 client usage.
+        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+
+        const auto& name = *it;
+        co_await backup_file(name);
+
+        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
+    }
+};
+
 future<> backup_task_impl::do_backup() {
     if (!co_await file_exists(_snapshot_dir.native())) {
         throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
     co_await process_snapshot_dir();
 
-    for (auto it = _files.begin(); it != _files.end() && !_ex; ++it) {
-        // Pre-upload break point. For testing abort in actual s3 client usage.
-        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
-
-        co_await backup_file(*it);
+    try {
+        while (!_sstable_comps.empty() && !_ex) {
+            auto ent = _sstable_comps.extract(_sstable_comps.begin());
+            co_await backup_sstable(ent.key(), ent.mapped());
+        }
 
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            _ex = impl::_as.abort_requested_exception_ptr();
-            break;
+        for (auto it = _non_sstable_files.begin(); it != _non_sstable_files.end() && !_ex; ++it) {
+            co_await backup_file(*it);
```

Comment:
For me this is more readable:
```suggestion
        for (auto& f : _non_sstable_files) {
            co_await backup_file(f);
            // we might encounter an exception in backup_file(), bail out
           if (_ex) { break; }
       }

Reply:
re-design in v5 ([a8bc344](https://github.com/scylladb/scylladb/commit/a8bc344156a32b48877ac1fa90736f07f9d77a6c))